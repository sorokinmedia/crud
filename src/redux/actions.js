const actions = {
	FETCH_CRUD_FILTER_VALUES: 'FETCH_CRUD_FILTER_VALUES',
	FETCH_CRUD_MODELS: 'FETCH_CRUD_MODELS',
	SET_CRUD_ACTIONS_FUNC: 'SET_CRUD_ACTIONS_FUNC',
	SET_CRUD_PARAMS: 'SET_CRUD_PARAMS',

	DELETE_MODEL: 'DELETE_MODEL',
	RESTORE_MODEL: 'RESTORE_MODEL',
	CHANGE_MODEL: 'CHANGE_MODEL',
	CREATE_MODEL: 'CREATE_MODEL',
	TOGGLE_CREATE_MODEL_MODAL: 'TOGGLE_CREATE_MODEL_MODAL',
	SET_MODEL_MODAL_FORM: 'SET_MODEL_MODAL_FORM',
	FETCH_CRUD_CHILDREN: 'FETCH_CRUD_CHILDREN',
	SET_UPLOADER_FILES: 'SET_UPLOADER_FILES',
	SET_UPLOADER_DEFAULT_FILE_LIST: 'SET_UPLOADER_DEFAULT_FILE_LIST',
	FETCH_FILE_CONFIG: 'FETCH_FILE_CONFIG',
	SET_CRUD_COLUMNS: 'SET_CRUD_COLUMNS',
	setCrudColumns: (columns, modelName) => ({ type: actions.SET_CRUD_COLUMNS, payload: { columns, modelName } }),
	fetchFileConfig: (url, modelName) => ({ type: actions.FETCH_FILE_CONFIG, payload: { url, modelName } }),
	setUploaderDefaultFileList: (defaultFileList, modelName) => ({
		type: actions.SET_UPLOADER_DEFAULT_FILE_LIST,
		payload: { defaultFileList, modelName }
	}),
	setUploaderFiles: (files, modelName) => ({ type: actions.SET_UPLOADER_FILES, payload: { files, modelName } }),
	fetchCrudChildren: (id, modelName, url) => ({
		type: actions.FETCH_CRUD_CHILDREN,
		payload: {
			id, params: { modelName }, url
		}
	}),
	toggleCreateModelModal: modelName => ({ type: actions.TOGGLE_CREATE_MODEL_MODAL, payload: { modelName } }),

	deleteModel: (id, action, modelName) => ({
		type: actions.DELETE_MODEL,
		payload: {
			id, action, modelName
		}
	}),
	restoreModel: (id, url, modelName) => ({
		type: actions.RESTORE_MODEL,
		payload: {
			id, url, modelName
		}
	}),
	createModel: (form, url, modelName) => ({
		type: actions.CREATE_MODEL,
		payload: {
			form, url, modelName
		}
	}),

	changeModel: (form, action, modelName) => ({
		type: actions.CHANGE_MODEL,
		payload: {
			form, action, modelName
		}
	}),

	setModelModalForm: (modalType, initialValues, action) => ({
		type: actions.SET_MODEL_MODAL_FORM,
		payload: {
			modalType, initialValues, action
		}
	}),

	setCrudActionsFunc: (func, modelName) => ({ type: actions.SET_CRUD_ACTIONS_FUNC, payload: { func, modelName } }),
	setCrudParams: params => ({ type: actions.SET_CRUD_PARAMS, payload: params }),
	fetchCrudModels: (params, filters) => ({ type: actions.FETCH_CRUD_MODELS, payload: { params, filters } }),
	fetchCrudFilterValues: (modelName, filter, query) => ({
		type: actions.FETCH_CRUD_FILTER_VALUES,
		payload: {
			query, filter, modelName
		}
	}),
};

export default actions;
