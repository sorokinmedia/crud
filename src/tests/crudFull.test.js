import { configure, mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16/build/index'
import React from 'react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import jsdom from 'jsdom'
import CrudFull, {CrudFull as CrudFullClass} from '../components/crud/crudFull'
import thunk from 'redux-thunk'
import { shallowToJson } from 'enzyme-to-json'
import { BrowserRouter } from 'react-router-dom'
import CreateModalForm from '../components/crud/create/createModelPopup'
import CrudView from '../components/crud/view/crudView'
import actions from '../redux/actions'
import responseMock from './responseMock'

const toggleCreateModelModal = jest.fn();
const { JSDOM } = jsdom;
const doc = new JSDOM('<!doctype html><html><body></body></html>');
global.document = doc;
global.window = {...doc.defaultView, confirm: jest.fn(() => true)};
global.navigator = { userAgent: 'node.js', };

// const {SignUpClass} = require('../../../containers/Page/signup');

configure({ adapter: new Adapter() });

const mockStore = configureStore([thunk]);

describe('it should render CrudFull', () => {
	let container;

	beforeAll(() => {
		container = shallow(<CrudFull store={mockStore({})} />);
	});

	it('it should render CrudFull connected container', () => {
		expect(shallowToJson(container)).toMatchSnapshot();
	});
});

describe('should mount CrudFull', () => {
	let container, store;
	const modelName = 'model';
	const crudActionsFunc = jest.fn();

	beforeAll(() => {
		store = mockStore({
			form: {},
			crudActionsFunc: { [modelName]: crudActionsFunc },
			crudModels: { [modelName]: responseMock },
			crudFilterValues: { [modelName]: {} },
			crudColumns: { [modelName]: responseMock.data.columns.map(e => ({ ...e, visible: true })) },
			isModalOpen: true
		});
		container = mount(<Provider store={store}>
			<BrowserRouter>
				<CrudFull
					crudRead="url"
					modelName="model"
					createDisabled={false}
				/>
			</BrowserRouter>
		</Provider>);
	});

	it('it should contains creatButton', () => {
		const button = container.find('button').filterWhere(btn => btn.props().name === 'createButton');
		expect(button.length).toBe(1);
	});

	it('it should contains crudView', () => {
		expect(container.find(CrudView).length).toBe(1);
	});

	it('should dispatch TOGGLE_CREATE_MODEL_MODAL by click', () => {
		const expectedAction = { type: actions.TOGGLE_CREATE_MODEL_MODAL };
		const button = container.find('button').filterWhere(btn => btn.props().name === 'createButton').first();

		button.simulate('click');
		const searched = store.getActions().find(e => e.type === expectedAction.type);

		expect(searched).toEqual(expectedAction)
	});

	it('it should contains crud modal', () => {
		const wrp = shallow(<CrudFullClass
			store={store}
			crudRead="url"
			modelName="model"
			createDisabled={false}
			setCrudActionsFunc={jest.fn()}
			setCrudParams={jest.fn()}
			filteredColumns={{}}
			isModalOpen={true}
			objectModal={{}}
		/>);
		expect(wrp.find(CreateModalForm).length).toBe(1);
	});
});
