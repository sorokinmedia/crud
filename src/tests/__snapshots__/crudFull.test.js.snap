// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render CrudFull it should render CrudFull connected container 1`] = `
<CrudFull
  CustomButtons={[Function]}
  btnStyle={Object {}}
  changeModel={[Function]}
  createButtonTitle="Добавить"
  createDisabled={true}
  createModel={[Function]}
  deleteModel={[Function]}
  iconTheme="outline"
  iconsProvider={[Function]}
  onDeleteConfirmMessageFunc={[Function]}
  restoreModel={[Function]}
  rowSelection={null}
  setCrudActionsFunc={[Function]}
  setCrudParams={[Function]}
  setModelModalForm={[Function]}
  size="default"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  submitShape={[Function]}
  tableProps={Object {}}
  tableStyle={Object {}}
  tableWrapper={null}
  toggleCreateModelModal={[Function]}
  updateShape={[Function]}
  uploadFilesSettings={null}
/>
`;
